app/services/turn_processor.rb

```diff
@@ -40,6 +40,50 @@ def attack_opponent
     else
       game.current_turn = 'player_1'
     end
+
+    if winner?
+      @messages << "Game over."
+    end
+
+  end
+
+  def winner?
+    game && player_1_winner? || game && player_2_winner?
+  end
+
+  def player_1_winner?
+    if player_2_board_total_hits >= player_2_total_occupied_spaces
+      winner_email(game.player_1)
+      return true
+    end
+  end
+
+  def player_2_winner?
+    if player_1_board_total_hits >= player_1_total_occupied_spaces
+      winner_email(game.player_2)
+      return true
+    end
+  end
+
+  def winner_email(player)
+    game.winner = User.find_by(api_key: player.api_key).email
+    game.save
+  end
+
+  def player_2_board_total_hits
+    game.player_2.board.board.flatten.map(&:values).flatten.map(&:contents).compact.select { |contents| contents.damage != 0 }.size
+  end
+
+  def player_2_total_occupied_spaces
+    game.player_2.board.board.flatten.map(&:values).flatten.map(&:contents).compact.size
+  end
+
+  def player_1_board_total_hits
+    game.player_1.board.board.flatten.map(&:values).flatten.map(&:contents).compact.select { |contents| contents.damage != 0 }.size
+  end
+
+  def player_1_total_occupied_spaces
+    game.player_1.board.board.flatten.map(&:values).flatten.map(&:contents).compact.size
   end
 
```

Comment:
This essentially compares a player's attacked spaces to the number of spaces occupied, for each player.

I do feel we should pull the winner methods into some module or something. It at first felt like it should be evauluated in turn processesor, but I don't like how long the class is getting.